src/hypothesis/stateful.py

```diff
@@ -426,18 +472,30 @@ def rules(cls):
 
         for k, v in inspect.getmembers(cls):
             r = getattr(v, RULE_MARKER, None)
-            while r is not None:
+            if r is not None:
                 cls.define_rule(
                     r.targets, r.function, r.arguments, r.precondition,
-                    r.parent_rule
                 )
-                r = r.parent_rule
         cls._rules_per_class[cls] = cls._base_rules_per_class.pop(cls, [])
         return cls._rules_per_class[cls]
 
     @classmethod
-    def define_rule(cls, targets, function, arguments, precondition=None,
-                    parent_rule=None):
+    def invariants(cls):
+        try:
+            return cls._invariants_per_class[cls]
```

Comment:
Given that this is more or less the same code as for rule I wonder if it would be worth factoring this out into a common method (it's up to you if you care enough to do this, I don't require it).